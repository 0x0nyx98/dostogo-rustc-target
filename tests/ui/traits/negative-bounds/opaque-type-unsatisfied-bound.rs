//@ compile-flags: -Znext-solver

#![feature(negative_bounds, negative_impls)]

trait Trait {}
impl !Trait for () {}

fn produce() -> impl !Trait {}
fn consume(_: impl Trait) {}

fn main() {
    consume(produce()); //~ ERROR the trait bound `impl !Trait: Trait` is not satisfied
}

fn weird0() -> impl Sized + !Sized {}
//~^ ERROR the size for values of type `()` cannot be known at compilation time [E0277]
//~| ERROR the size for values of type `()` cannot be known at compilation time [E0277]
//~| ERROR the size for values of type `impl !Sized + Sized` cannot be known at compilation time [E0277]
//~| ERROR the size for values of type `()` cannot be known at compilation time [E0277]
fn weird1() -> impl !Sized + Sized {}
//~^ ERROR the size for values of type `()` cannot be known at compilation time [E0277]
//~| ERROR the size for values of type `()` cannot be known at compilation time [E0277]
//~| ERROR the size for values of type `impl !Sized + Sized` cannot be known at compilation time [E0277]
//~| ERROR the size for values of type `()` cannot be known at compilation time [E0277]
fn weird2() -> impl !Sized {}
//~^ ERROR the size for values of type `()` cannot be known at compilation time [E0277]
//~| ERROR the size for values of type `()` cannot be known at compilation time [E0277]
//~| ERROR the size for values of type `impl !Sized` cannot be known at compilation time [E0277]
//~| ERROR the size for values of type `()` cannot be known at compilation time [E0277]
