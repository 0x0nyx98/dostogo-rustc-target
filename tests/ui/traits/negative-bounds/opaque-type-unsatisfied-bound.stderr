error[E0271]: type mismatch resolving `weird0::{opaque#0} normalizes-to _`
  --> $DIR/opaque-type-unsatisfied-bound.rs:15:16
   |
LL | fn weird0() -> impl Sized + !Sized {}
   |                ^^^^^^^^^^^^^^^^^^^ types differ

error[E0271]: type mismatch resolving `weird0::{opaque#0} normalizes-to _`
  --> $DIR/opaque-type-unsatisfied-bound.rs:15:36
   |
LL | fn weird0() -> impl Sized + !Sized {}
   |                                    ^^ types differ

error[E0277]: the size for values of type `impl !Sized + Sized` cannot be known at compilation time
  --> $DIR/opaque-type-unsatisfied-bound.rs:15:16
   |
LL | fn weird0() -> impl Sized + !Sized {}
   |                ^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `Sized` is not implemented for `impl !Sized + Sized`
   = note: the return type of a function must have a statically known size

error[E0271]: type mismatch resolving `weird0::{opaque#0} normalizes-to _`
  --> $DIR/opaque-type-unsatisfied-bound.rs:15:1
   |
LL | fn weird0() -> impl Sized + !Sized {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ types differ

error[E0271]: type mismatch resolving `weird1::{opaque#0} normalizes-to _`
  --> $DIR/opaque-type-unsatisfied-bound.rs:17:16
   |
LL | fn weird1() -> impl !Sized + Sized {}
   |                ^^^^^^^^^^^^^^^^^^^ types differ

error[E0271]: type mismatch resolving `weird1::{opaque#0} normalizes-to _`
  --> $DIR/opaque-type-unsatisfied-bound.rs:17:36
   |
LL | fn weird1() -> impl !Sized + Sized {}
   |                                    ^^ types differ

error[E0277]: the size for values of type `impl !Sized + Sized` cannot be known at compilation time
  --> $DIR/opaque-type-unsatisfied-bound.rs:17:16
   |
LL | fn weird1() -> impl !Sized + Sized {}
   |                ^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `Sized` is not implemented for `impl !Sized + Sized`
   = note: the return type of a function must have a statically known size

error[E0271]: type mismatch resolving `weird1::{opaque#0} normalizes-to _`
  --> $DIR/opaque-type-unsatisfied-bound.rs:17:1
   |
LL | fn weird1() -> impl !Sized + Sized {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ types differ

error[E0271]: type mismatch resolving `weird2::{opaque#0} normalizes-to _`
  --> $DIR/opaque-type-unsatisfied-bound.rs:19:16
   |
LL | fn weird2() -> impl !Sized {}
   |                ^^^^^^^^^^^ types differ

error[E0271]: type mismatch resolving `weird2::{opaque#0} normalizes-to _`
  --> $DIR/opaque-type-unsatisfied-bound.rs:19:28
   |
LL | fn weird2() -> impl !Sized {}
   |                            ^^ types differ

error[E0277]: the size for values of type `impl !Sized` cannot be known at compilation time
  --> $DIR/opaque-type-unsatisfied-bound.rs:19:16
   |
LL | fn weird2() -> impl !Sized {}
   |                ^^^^^^^^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `Sized` is not implemented for `impl !Sized`
   = note: the return type of a function must have a statically known size

error[E0271]: type mismatch resolving `weird2::{opaque#0} normalizes-to _`
  --> $DIR/opaque-type-unsatisfied-bound.rs:19:1
   |
LL | fn weird2() -> impl !Sized {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^ types differ

error[E0277]: the trait bound `impl !Trait: Trait` is not satisfied
  --> $DIR/opaque-type-unsatisfied-bound.rs:12:13
   |
LL |     consume(produce());
   |     ------- ^^^^^^^^^ the trait `Trait` is not implemented for `impl !Trait`
   |     |
   |     required by a bound introduced by this call
   |
note: required by a bound in `consume`
  --> $DIR/opaque-type-unsatisfied-bound.rs:9:20
   |
LL | fn consume(_: impl Trait) {}
   |                    ^^^^^ required by this bound in `consume`

error: aborting due to 13 previous errors

Some errors have detailed explanations: E0271, E0277.
For more information about an error, try `rustc --explain E0271`.
